libraries/chain/transaction_context.cpp

```diff
@@ -496,28 +523,26 @@ namespace eosio { namespace chain {
                         ("billed", billed_us)( "billable", objective_duration_limit.count() )
             );
          } else {
-            auto assert_msg =  [&](bool graylisted, bool subjective) {
-               std::string assert_msg =   "billed CPU time (${billed} us) is greater than the maximum";
-               assert_msg += graylisted ? " greylisted" : "";
-               assert_msg +=              " billable CPU time for the transaction (${billable} us)";
-               assert_msg += subjective ? " with a subjective cpu of (${subjective} us)" : "";
-               return assert_msg;
-            };
             auto graylisted = cpu_limit_due_to_greylist && cpu_limited_by_account;
-            auto subjective =  subjective_billed_us > 0;
             // exceeds trx.max_cpu_usage_ms or cfg.max_transaction_cpu_usage if objective_duration_limit is greater
-            auto limit = graylisted ? account_cpu_limit : (cpu_limited_by_account ? account_cpu_limit : objective_duration_limit.count()); 
-           
-            if (graylisted)
-               EOS_ASSERT( billed_us <= limit,
-                           greylist_cpu_usage_exceeded, 
-                           assert_msg(graylisted, subjective),
-                           ("billed", billed_us)("billable", limit)("subjective", subjective_billed_us));
-            else
-               EOS_ASSERT( billed_us <= limit,
-                           tx_cpu_usage_exceeded,
-                           assert_msg(graylisted, subjective), 
-                           ("billed", billed_us)("billable", limit)("subjective", subjective_billed_us));           
+            auto account_limit = graylisted ? account_cpu_limit : (cpu_limited_by_account ? account_cpu_limit : objective_duration_limit.count());
+
+            if( billed_us > account_limit ) {
+               fc::microseconds tx_limit;
+               std::string assert_msg = "billed CPU time (${billed} us) is greater than the maximum";
+               assert_msg += graylisted ? " greylisted" : "";
+               assert_msg += " billable CPU time for the transaction (${billable} us)";
+               assert_msg += subjective_billed_us > 0 ? " with a subjective cpu of (${subjective} us)" : "";
+               assert_msg += get_tx_cpu_usage_exceeded_reason_msg( tx_limit );
```

Comment:
I think I would declare the `std::string assert_msg` and do `assert_msg.reserve(1024)` to avoid a new memory allocation at each `+=` operator. Same in `validate_account_cpu_usage_estimate` below.

Reply:
Done.